master/custom/release_dashboard.py

```diff
@@ -1,122 +1,730 @@
+            # any info (which we'll also get when information is
+            # scrubbed after some months)
+            return []
+        return [Change(self, info) for info in infos]
+
+    @cached_property
+    def started_at(self):
+        if self["started_at"]:
+            return datetime.datetime.fromtimestamp(self["started_at"],
+                                                   tz=datetime.timezone.utc)
 
-    @release_status_app.route("/index.html")
-    def main():
-        nonlocal cache
+    @cached_property
+    def age(self):
+        if self["started_at"]:
+            return self._root.now - self.started_at
 
-        force_refresh = request.args.get("refresh", "").lower() in {"1", "yes", "true"}
+    @cached_property
+    def results_symbol(self):
+        if self["results"] == buildbot.process.results.FAILURE:
+            return '\N{HEAVY BALLOT X}'
+        if self["results"] == buildbot.process.results.WARNINGS:
+            return '\N{WARNING SIGN}'
+        if self["results"] == buildbot.process.results.SUCCESS:
+            return '\N{HEAVY CHECK MARK}'
+        if self["results"] == buildbot.process.results.SKIPPED:
+            return '\N{CIRCLED MINUS}'
+        if self["results"] == buildbot.process.results.EXCEPTION:
+            return '\N{CIRCLED DIVISION SLASH}'
+        if self["results"] == buildbot.process.results.RETRY:
+            return '\N{ANTICLOCKWISE OPEN CIRCLE ARROW}'
+        if self["results"] == buildbot.process.results.CANCELLED:
+            return '\N{CIRCLED TIMES}'
+        return str(self["results"])
 
-        if cache is not None and not force_refresh:
-            result, deadline = cache
-            if time.monotonic() <= deadline:
-                return result
+    @cached_property
+    def results_string(self):
+        return buildbot.process.results.statusToString(self["results"])
 
-        result = get_release_status()
-        deadline = time.monotonic() + CACHE_DURATION
-        cache = (result, deadline)
+    @cached_property
+    def css_color_class(self):
+        if self["results"] == buildbot.process.results.SUCCESS:
+            return 'success'
+        if self["results"] == buildbot.process.results.WARNINGS:
+            return 'warning'
+        if self["results"] == buildbot.process.results.FAILURE:
+            return 'danger'
+        return 'unknown'
+
+    @cached_property
+    def junit_results(self):
+        filepath = (
+            self._root._app.test_result_dir
+            / self.builder.branch.tag
+            / self.builder["name"]
+            / f'build_{self["number"]}.xml'
+        )
+        try:
+            file = filepath.open()
+        except OSError:
+            return None
+        with file:
+            etree = ElementTree.parse(file)
+        result = JunitResult(self, {})
+        for element in etree.iterfind('.//error/..'):
+            result.add(element)
         return result
```

Comment:
(This assumes there is a logger somewhere, but you can drop that)

```suggestion
@cached_property
def junit_results(self):
    if not self._root._app.test_result_dir:
        return None
        
    try:
        filepath = (
            self._root._app.test_result_dir
            / self.builder.branch.tag
            / self.builder["name"]
            / f'build_{self["number"]}.xml'
        ).resolve()
        
        # Ensure path doesn't escape test_result_dir
        if not str(filepath).startswith(str(self._root._app.test_result_dir)):
            logger.warning(f"Attempted path traversal: {filepath}")
            return None
            
        if not filepath.is_file():
            return None
            
        with filepath.open() as file:
            etree = ElementTree.parse(file)
            
    except OSError as e:
        logger.error(f"Failed to read JUnit results: {e}")
        return None
    except ElementTree.ParseError as e:
        logger.error(f"Failed to parse JUnit XML: {e}")
        raise XMLParseError(f"Malformed JUnit XML: {e}") from e
```